src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+
+                                resultOrSet.Add(resultITuplePattern, syntax);
+                            }
+                        }
+                    }
+                }
+
+                return result;
+            }
+
+
+            // Update `pattern1 or pattern2 or ... or patternN` tree, but with the i-th pattern/leaf substituted.
+            // We return either an updated node or a count of leaves found in that node.
+            static (BoundBinaryPattern? updated, int leafCount) updateBinaryOrTree(BoundBinaryPattern binaryPattern, int leafIndex, BoundPattern pattern)
+            {
+                int leftLeafCount;
+                if (binaryPattern.Left is BoundBinaryPattern { Disjunction: true } left)
+                {
+                    (var updatedLeft, leftLeafCount) = updateBinaryOrTree(left, leafIndex, pattern);
+                    if (updatedLeft is not null)
+                    {
+                        // The index we are looking for is on the left
+                        return (binaryPattern.WithLeft(updatedLeft), -1);
+                    }
+                }
+                else
+                {
+                    // Reached a leaf on the left
+                    leftLeafCount = 1;
+                    if (leafIndex == 0)
+                    {
+                        // The index we are looking for is on the left
+                        return (binaryPattern.WithLeft(pattern), -1);
+                    }
+                }
+
+                if (binaryPattern.Right is BoundBinaryPattern { Disjunction: true } right)
+                {
+                    var (updatedRight, rightLeafCount) = updateBinaryOrTree(right, leafIndex - leftLeafCount, pattern);
+                    if (updatedRight is not null)
+                    {
+                        return (binaryPattern.WithRight(updatedRight), -1);
+                    }
+
+                    return (null, leftLeafCount + rightLeafCount);
+                }
+
+                if (leafIndex == leftLeafCount)
+                {
+                    // The index we are looking for is on the right
+                    return (binaryPattern.WithRight(pattern), -1);
+                }
+
+                return (null, leftLeafCount + 1);
+            }
+
+            static void addPatternsFromOrTree(BoundPattern pattern, ArrayBuilder<BoundPattern> builder)
+            {
+                if (pattern is BoundBinaryPattern { Disjunction: true } orPattern)
+                {
+                    addPatternsFromOrTree(orPattern.Left, builder);
+                    addPatternsFromOrTree(orPattern.Right, builder);
+                }
+                else
+                {
+                    builder.Add(pattern);
+                }
+            }
+        }
+
+        /// <summary>
+        /// When there are multiple `or` patterns, such as `(A or B) and (C or D)`
+        /// we'll expand then two sets:
+        /// 1. `A` and `B`
```

Comment:
>`A` and `B` [](http://example.com/codeflow?start=15&length=11)

Here and in other similar paces, consider placing cases on different lines, the "and" word is is easily confused with an OperatorToken in an `and` binary pattern